pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,628 @@
+		case <-c.localNodeEventCh:
+			c.syncLocalNodeStatus()
+		case hashEvent := <-c.conHashEventCh:
+			c.syncConHash(&hashEvent)
+		}
+	}
+}
+
+//func (c *Cluster) localworker(stopCh <-chan struct{}) {
+//	for {
+//		select {
+//		case <-stopCh:
+//			return
+//		case hashEvent := <-c.conHashEventCh:
+//			c.syncConHash(&hashEvent)
+//		}
+//	}
+//}
+
+func (c *Cluster) syncLocalNodeStatus() {
+	shouldJoin := c.shouldJoinCluster(c.nodeName)
+	joined := c.LocalNodeJoined()
+	if joined && !shouldJoin {
+		// filter nodes through nodeSelector from externalIPPools
+		otherClusterNodes := c.syncClusterMembers()
+		c.addExistingMembers(otherClusterNodes...)
+		klog.InfoS("Received a join signal, local node will join cluster", "node name", c.nodeName)
+		c.joinOrLeaveCluster(false)
+	} else if !joined && shouldJoin {
+		klog.InfoS("Received a leave signal, local node will leave cluster", "node name", c.nodeName)
+		c.joinOrLeaveCluster(false)
+	}
+	return
+}
+
+// if local node alive
+// 1.add ippool, update one key
+// 2.update ipPool labels, nodes matched changed, update one key, node need to
+// 3.delete ippool, del one key
+// add node, node will join and then we can receive a node join signal, no need to handle
+// update node, labels changes and maybe all key should change
+// delete node, node will leave and no need to handle
+func (c *Cluster) syncConHash(event *clusterHashEvent) {
+	// only alive node was in cluster and need to holds the consist hashMap
+	if !c.LocalNodeJoined() {
+		return
+	}
+	c.conHashMapRWLock.Lock()
+	defer c.conHashMapRWLock.Unlock()
+	switch eventType, obj := event.eventType, event.obj; eventType {
+	case ipPoolCreate:
+		// if IPPool create, create a new conHash, add alive node keys, no need to handle egresses
+		ipPool := obj.(*v1alpha2.ExternalIPPool)
+		c.syncConHashByIPPool(ipPool)
+	case ipPoolUpdate:
+		// if IPPool update, alive nodes may leave or join, enqueue egresses
+		ipPool := obj.(*v1alpha2.ExternalIPPool)
+		c.syncConHashByIPPool(ipPool)
+		c.notify(ipPool.Name, NotifyEventTypeIPPoolChanged)
+	case ipPoolDelete:
+		// if IPPool delete, all the egresses has deleted, refs https://github.com/antrea-io/antrea/pull/2237
+		// just delete the conHash
+		ipPool := obj.(*v1alpha2.ExternalIPPool)
+		delete(c.conHashMap, ipPool.Name)
+	case nodeLabelUpdate:
+		// node alive in cluster, and may leave from a IPPool or join a IPPool
+		// all egresses with a IPPool would enqueue
+		poolNames := obj.(sets.String)
+		for _, poolName := range poolNames.List() {
+			ipPool, _ := c.ipPoolLister.Get(poolName)
+			c.syncConHashByIPPool(ipPool)
+			c.notify(ipPool.Name, NotifyEventTypeIPPoolChanged)
+		}
+	}
+}
+
+func (c *Cluster) syncConHashByIPPool(ipPool *v1alpha2.ExternalIPPool) {
+	nodeSel, _ := metav1.LabelSelectorAsSelector(&ipPool.Spec.NodeSelector)
+	nodes, _ := c.nodeLister.List(nodeSel)
+	conHash, ok := c.conHashMap[ipPool.Name]
+	if !ok {
+		conHash = newNodeConsistentHashMap()
```

Comment:
Can it work? I think the consistentHash will always expand and never shrink if it only creates new map when it's missing.

Reply:
I think it has fixed in new version